server/src/main/java/org/opensearch/ingest/IngestService.java

```diff
@@ -802,69 +822,87 @@ private void runBulkRequestInBatch(
-                pipelinesInfoList.add(new IngestPipelineInfo(pipelineId, IngestPipelineType.DEFAULT));
-            }
-            if (IngestService.NOOP_PIPELINE_NAME.equals(finalPipelineId) == false) {
-                pipelinesInfoList.add(new IngestPipelineInfo(finalPipelineId, IngestPipelineType.FINAL));
-            }
+                List<IngestPipelineInfo> pipelinesInfoList = new ArrayList<>();
 
-            if (IngestService.NOOP_PIPELINE_NAME.equals(systemPipelineId) == false) {
-                pipelinesInfoList.add(new IngestPipelineInfo(systemPipelineId, IngestPipelineType.SYSTEM_FINAL));
-            }
+                if (pipelineId != null && IngestService.NOOP_PIPELINE_NAME.equals(pipelineId) == false) {
+                    pipelinesInfoList.add(new IngestPipelineInfo(pipelineId, IngestPipelineType.DEFAULT));
+                }
+                if (finalPipelineId != null && IngestService.NOOP_PIPELINE_NAME.equals(finalPipelineId) == false) {
+                    pipelinesInfoList.add(new IngestPipelineInfo(finalPipelineId, IngestPipelineType.FINAL));
+                }
+                if (systemPipelineId != null && IngestService.NOOP_PIPELINE_NAME.equals(systemPipelineId) == false) {
+                    pipelinesInfoList.add(new IngestPipelineInfo(systemPipelineId, IngestPipelineType.SYSTEM_FINAL));
+                }
 
-            if (pipelinesInfoList.isEmpty()) {
-                if (counter.decrementAndGet() == 0) {
-                    onCompletion.accept(originalThread, null);
+                // Create index request wrappers for all index requests with pipelines that will be executed
+                if (pipelinesInfoList.isEmpty() == false) {
+                    indexRequestWrappers.add(new IndexRequestWrapper(slot, innerSlot, indexRequest, actionRequest, pipelinesInfoList));
                 }
-                assert counter.get() >= 0;
-            } else {
-                indexRequestWrappers.add(new IndexRequestWrapper(i, indexRequest, actionRequest, pipelinesInfoList));
             }
 
-            i++;
+            slot++;
+        }
+
+        // If we don't need to execute any pipelines, we exit pipeline execution early
+        if (indexRequestWrappers.isEmpty()) {
+            onCompletion.accept(originalThread, null);
+            return;
         }
 
+        // Used to track progress of pipeline executions on index requests across parallel threads
+        // based on the total number of index request wrappers generated
+        final AtomicInteger totalIndexRequestCounter = new AtomicInteger(indexRequestWrappers.size());
+
         int batchSize = numberOfActionRequests;
         List<List<IndexRequestWrapper>> batches = prepareBatches(batchSize, indexRequestWrappers);
         logger.debug("batchSize: {}, batches: {}", batchSize, batches.size());
 
         for (List<IndexRequestWrapper> batch : batches) {
             executePipelinesInBatchRequests(
```

Comment:
Can we explore on running these request concurrently using executor service?

Reply:
Can you elaborate? What do you mean run the request concurrently using executor service?

Reply:
Ah I see what you mean, yes, pipeline execution is done parallelized and async already. The `executePipelinesInBatchRequests` call is sync but it goes async at the `innerBatchExecute` call [here](https://github.com/opensearch-project/OpenSearch/blob/40dfbd39cfb5cf00546fdb00e0fb8053dc5b1efb/server/src/main/java/org/opensearch/ingest/IngestService.java#L967) implemented in this RFC: https://github.com/opensearch-project/OpenSearch/issues/12457